api/src/adapters/db/type_decorators/postgres_type_decorators.py

```diff
@@ -1,74 +1,48 @@
-from enum import StrEnum
+from enum import IntEnum, StrEnum
 from typing import Any, Type
 
-from sqlalchemy import Text
+from sqlalchemy import Integer
 from sqlalchemy.types import TypeDecorator
 
+from src.db.models.lookup import LookupRegistry, LookupTable
 
-class StrEnumColumn(TypeDecorator):
-    """
-    This class handles converting StrEnum objects into strings when writing to the DB,
-    and converting those strings back into the provided StrEnum when reading from the DB.
-
-    Example Usage::
-
-        from enum import StrEnum
-        from sqlalchemy.orm import Mapped, mapped_column
-        from src.db.models.base import Base, IdMixin, TimestampMixin
-
-        # Define a StrEnum somewhere
-        class ExampleEnum(StrEnum):
-            VALUE_A = "a"
-            VALUE_B = "b"
-            VALUE_C = "c"
-
-        # Create your DB model, specifying the column type like so
-        class Example(Base, IdMixin, TimestampMixin):
-            __tablename__ = "example"
-
-            example_column: Mapped[ExampleEnum] = mapped_column(StrEnumColumn(ExampleEnum))
-            ...
-
-
-        # Use the model - when the value is written to the DB, just "a" will be written
-        example = Example(example_column=ExampleEnum.VALUE_A)
 
+class LookupColumn(TypeDecorator):
+    """
+    A Postgres column decorator that wraps
+    an integer column representing a lookup int.
 
-    See: https://docs.sqlalchemy.org/en/20/core/custom_types.html#types-typedecorator
+    This takes in the LookupTable that the lookup value
+    is stored in, and handles converting the Lookup object
+    in-code into the integer in the DB automatically (and the reverse).
     """
 
-    impl = Text
+    impl = Integer
 
     cache_ok = True
 
-    def __init__(self, lookup_enum: Type[StrEnum], *args: Any, **kwargs: Any):
+    def __init__(self, lookup_table: Type[LookupTable], *args: Any, **kwargs: Any):
         super().__init__(*args, **kwargs)
-        self.lookup_enum = lookup_enum
+        self.lookup_table = lookup_table
 
-    def process_bind_param(self, value: Any | None, dialect: Any) -> str | None:
-        """
-        Method for converting a StrEnum when writing TO the DB
-        """
+    def process_bind_param(self, value: Any | None, dialect: Any) -> int | None:
         if value is None:
             return None
 
-        if not isinstance(value, self.lookup_enum):
+        if not isinstance(value, (StrEnum, IntEnum)):
```

Comment:
An idea for making this check stronger (currently it could allow the wrong enum type to be set I think).

- Add an `enum_type` argument to `__init__()` and store it as a member variable.
- Change this line to `isinstance(value, self.enum_type)`
- Wherever `LookupColumn` is created, add the enum type. For example:

```python
   category: Mapped[OpportunityCategory | None] = mapped_column(
        "opportunity_category_id",
        LookupColumn(LkOpportunityCategory, OpportunityCategory),  # 2nd argument added
        ForeignKey(LkOpportunityCategory.opportunity_category_id),
        index=True,
    )
```

Reply:
I could also just make it so each `Lk___` class has the enum specifically registered with it and use that? It registers the mapping of enum values, but if I just add the enum itself that would accomplish the same thing

Reply:
In the LookupConfig we already were grabbing all of the enums that were associated with a config (note that while not used now - this is setup to allow multiple python enums for a given Lk table), just piped that through via the LookupRegistry